sound/soc/sof/probe.c

```diff
@@ -0,0 +1,205 @@
+// SPDX-License-Identifier: (GPL-2.0 OR BSD-3-Clause)
+//
+// This file is provided under a dual BSD/GPLv2 license.  When using or
+// redistributing this file, you may do so under either license.
+//
+// Copyright(c) 2019 Intel Corporation. All rights reserved.
+//
+// Author: Cezary Rojewski <cezary.rojewski@intel.com>
+//
+
+#include "sof-priv.h"
+#include "probe.h"
+
+int sof_ipc_probe_init(struct snd_sof_dev *sdev,
+		size_t buffer_size)
+{
+	struct sof_ipc_probe_dma_set_params *msg;
+	struct sof_ipc_reply reply;
+	unsigned int stream_tag = INVALID_NODE_ID;
+	size_t size;
+	int ret;
+
+	size = struct_size(msg, dma, 1);
+	msg = kzalloc(size, GFP_KERNEL);
+	if (!msg)
+		return -ENOMEM;
+
+	if (sdev->extractor)
+		stream_tag = sdev->extractor;
+
+	msg->hdr.size = size;
+	msg->hdr.cmd = SOF_IPC_GLB_PROBE | SOF_IPC_PROBE_INIT;
+	msg->dma[0].stream_tag = stream_tag;
+	msg->dma[0].dma_buffer_size = buffer_size;
+
+	ret = sof_ipc_tx_message(sdev->ipc, msg->hdr.cmd, msg, msg->hdr.size,
+			&reply, sizeof(reply));
+	kfree(msg);
+	return ret;
+}
+EXPORT_SYMBOL(sof_ipc_probe_init);
+
+int sof_ipc_probe_deinit(struct snd_sof_dev *sdev)
+{
+	struct sof_ipc_cmd_hdr msg;
+	struct sof_ipc_reply reply;
+
+	msg.size = sizeof(msg);
+	msg.cmd = SOF_IPC_GLB_PROBE | SOF_IPC_PROBE_DEINIT;
+
+	return sof_ipc_tx_message(sdev->ipc, msg.cmd, &msg, msg.size,
+			&reply, sizeof(reply));
+}
+EXPORT_SYMBOL(sof_ipc_probe_deinit);
+
+static int sof_ipc_probe_get(struct snd_sof_dev *sdev, unsigned int cmd,
+		void **params, size_t *bytes)
+{
+	struct sof_ipc_probe_get_params msg = {0};
+	struct sof_ipc_probe_get_params *reply;
```

Comment:
@crojewsk just curious, why do you use a pointer for reply but not msg?

Reply:
reply is the one carrying payload from FW. Since the side is unknown upfront, allocation of buffer with size==MAILBOX_SIZE is the expected behavior of host from FW's perspective.
Please note reply local is of type: sof_ipc_probe_get_params which contains union field consisting of two flex arrays.